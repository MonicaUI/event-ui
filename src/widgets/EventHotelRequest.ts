import { connect } from 'react-redux';
import { withLoading } from 'nucleus-guestside-site/src/redux/modules/dialogContainer';
import HotelRequestWidget from 'event-widgets/lib/HotelRequest/HotelRequestWidget';
import {
  toggleOwnAccommodation,
  showHotelRequestSummaryView,
  setAnotherHotelRequestFlag,
  updateExpandedHotels,
  loadRoommateData,
  searchRoommates
} from '../redux/travelCart/index';
import { saveHotelRoomRequests, openHotelRequestDeleteConfirmation } from '../redux/travelCart/workflow';
import * as currentRegistrant from '../redux/selectors/currentRegistrant';
import {
  getHotelRoomBookingsToDisplay,
  getAttendeeRegTypeAndAdmItemIds,
  getSelectedRoomsInHotelRequests,
  createDerivedRoomRegTypeAssociations
} from '../utils/travelUtils';
import * as travel from 'event-widgets/redux/selectors/eventTravel';
import { getActiveAccountHotelBillingInstructions, areRegistrationActionsDisabled } from '../redux/selectors/shared';
import { openShoulderDateApprovalWarningDialog } from '../dialogs';
import { transformEventDates } from 'event-widgets/utils/travelUtils';
import { getRegistrationPathIdOrDefault } from '../redux/selectors/currentRegistrationPath';
import { getIn } from 'icepick';
import { getRegisteredAdmissionItemForPrimaryAndGuests } from '../redux/selectors/currentRegistrant';
import getDialogContainerStyle from '../dialogs/shared/getDialogContainerStyle';

export default connect(
  (state: $TSFixMe) => {
    const primaryEventRegistration = currentRegistrant.getEventRegistration(state) || {};
    const guestsEventRegistrations = currentRegistrant.getConfirmedGuests(state) || [];
    const isGroupRegistration = currentRegistrant.isGroupRegistration(state);

    const hotelRequestsForCurrentInvitee = getHotelRoomBookingsToDisplay(
      state,
      primaryEventRegistration.eventRegistrationId,
      guestsEventRegistrations.map(r => r.eventRegistrationId)
    );

    const hotelRegRules = travel.getHotelRegRules(state);
    const admissionItem = currentRegistrant.getSelectedAdmissionItemDefinition(state) || {};
    const isPlanner = state.defaultUserSession.isPlanner;
    const isRegMod = currentRegistrant.isRegistrationModification(state);

    const isPendingInvitee = currentRegistrant.isRegApprovalRequired(state);
    const isPendingInviteeModification = isPendingInvitee && isRegMod;

    const isRegistrationActionDisabled = areRegistrationActionsDisabled(state);

    // show planner fields only if planner is registering OR (modifying an approved invitee).
    const displayPlannerFields = isPlanner && !isPendingInviteeModification;
    const currentRegistrationPathId = getRegistrationPathIdOrDefault(state);
    const roommateRequestSettings = getIn(state, [
      'appData',
      'registrationSettings',
      'registrationPaths',
      currentRegistrationPathId,
      'travelSettings',
      'hotelRequestSettings',
      'roommateRequestSettings'
    ]);
    const hotelRequestSessionInfo = getIn(state, ['travelCart', 'userSession', 'hotelRequest']);
    const allowedAttendeeTypes = getIn(state, [
      'appData',
      'registrationSettings',
      'registrationPaths',
      currentRegistrationPathId,
      'travelSettings',
      'hotelRequestSettings',
      'allowedAttendeeTypes'
    ]);
    const isRoomVisibilityExperimentEnabled = getIn(state, ['experiments', 'isHotelRoomVisibilityExperimentEnabled']);
    const attendeeRegTypeAndAdmItemIds = getAttendeeRegTypeAndAdmItemIds(
      travel.getHotelVisibilityOption(state),
      currentRegistrant.getRegistrationTypesForPrimaryAndGuests(state, allowedAttendeeTypes),
      getRegisteredAdmissionItemForPrimaryAndGuests(state, allowedAttendeeTypes),
      isRoomVisibilityExperimentEnabled
    );
    const hotels = travel.getHotels(state);
    const regPaths = getIn(state, ['appData', 'registrationSettings', 'registrationPaths']);
    const derivedRoomRegTypeAssociations = createDerivedRoomRegTypeAssociations(
      state,
      travel.getHotelVisibilityOption(state),
      isRoomVisibilityExperimentEnabled,
      hotels,
      regPaths
    );
    const selectedRooms = getSelectedRoomsInHotelRequests(
      hotelRequestsForCurrentInvitee,
      hotels,
      isPlanner,
      travel.getHotelVisibilityOption(state),
      currentRegistrationPathId,
      attendeeRegTypeAndAdmItemIds,
      isRoomVisibilityExperimentEnabled,
      derivedRoomRegTypeAssociations
    );

    const isSummaryView =
      (selectedRooms.length > 0 &&
        hotelRequestSessionInfo.expandedHotels.length === 0 &&
        !hotelRequestSessionInfo.isMakingAnotherRequest) ||
      hotelRequestSessionInfo.isSummaryView;

    return {
      hotelsData: {
        ...state.eventTravel.hotelsData,
        selectedRooms,
        isOwnAccommodation: hotelRequestSessionInfo.ownBooking
      },
      translateCurrency: state.text.resolver.currency,
      translateDate: state.text.translateDate,
      event: transformEventDates(state),
      registrationPathId: getRegistrationPathIdOrDefault(state),
      registrationTypeId: currentRegistrant.getRegistrationTypeId(state),
      attendeeRegTypeAndAdmItemIds,
      isPlanner,
      isGroupRegistration,
      isSiteEditorPreview: false,
      roommateRequestSettings,
      registrations: {
        primary: primaryEventRegistration,
        guests: guestsEventRegistrations
      },
      roommates: state.travelCart.roommates,
      hotelRegRules,
      admissionItemId: admissionItem.id,
      displayPlannerFields,
      accountHotelBillingInstructions: getActiveAccountHotelBillingInstructions(state),
      isRegMod,
      isSummaryView,
      expandedHotels: hotelRequestSessionInfo.expandedHotels,
      isPendingInvitee,
      isRoomVisibilityExperimentEnabled,
      derivedRoomRegTypeAssociations,
      dialogContainerStyle: getDialogContainerStyle(state),
      isRegistrationActionDisabled
    };
  },
  {
    toggleOwnAccommodation,
    showHotelRequestSummaryView,
    updateExpandedHotels,
    setAnotherHotelRequestFlag,
    onRoommateCriteriaChange: withLoading(loadRoommateData),
    onRoommateSearch: searchRoommates,
    onSubmitRoomRequest: saveHotelRoomRequests,
    onClearRoomSelection: openHotelRequestDeleteConfirmation,
    onRequestShoulderDatesRequiringApproval: openShoulderDateApprovalWarningDialog
  }
)(HotelRequestWidget);
