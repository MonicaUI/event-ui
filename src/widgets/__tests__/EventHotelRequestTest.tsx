import React from 'react';
import { shallow } from 'enzyme/build';
import EventHotelRequest from '../EventHotelRequest';
import EventSnapshot from '../../../fixtures/EventSnapshot.json';

jest.mock('event-widgets/utils/travelUtils', () => {
  return {
    transformEventDates: state => ({
      ...state.event,
      timezone: {
        id: 35,
        name: 'Eastern Time',
        nameResourceKey: 'Event_Timezone_Name_35__resx',
        plannerDisplayName: '(GMT-05:00) Eastern [US & Canada]',
        hasDst: true,
        utcOffset: -300,
        abbreviation: 'ET',
        abbreviationResourceKey: 'Event_Timezone_Abbr_35__resx',
        dstInfo: []
      }
    })
  };
});

let isHotelRequestEnabled = true;

function getState() {
  return {
    environment: 'S437',
    event: {
      timezone: 35,
      products: {
        admissionItems: {
          '3349dc98-e303-4bf9-b698-e1bdae85f946': {
            id: '3349dc98-e303-4bf9-b698-e1bdae85f946'
          }
        }
      },
      registrationTypes: {
        'b8e50dbf-7437-4eec-a964-bc27d79d0372': {
          id: 'b8e50dbf-7437-4eec-a964-bc27d79d0372'
        }
      }
    },
    registrationForm: {
      regCart: {
        eventRegistrations: {
          123456789: {
            eventRegistrationId: '123456789',
            registrationTypeId: 'b8e50dbf-7437-4eec-a964-bc27d79d0372',
            registrationPathId: 'b8e50dbf-7437-4eec-a964-bc27d79d0373',
            productRegistrations: [
              {
                productId: '3349dc98-e303-4bf9-b698-e1bdae85f946',
                productType: 'AdmissionItem',
                requestedAction: 'REGISTER'
              }
            ]
          }
        }
      }
    },

    eventTravel: {
      hotelsData: {
        hotels: [
          {
            id: '5ea37a1e-299a-4dc9-b1bf-c0ef74379fe2',
            name: 'aa',
            code: '',
            description: '',
            websiteLink: '',
            phone: '',
            fax: '',
            address: '',
            isActive: true,
            hotelId: '9d4f772f-dd54-4e56-a436-d91aab911c9c',
            registrationLinkUrl: '',
            rateCode: '',
            cancellationPolicy: '',
            arrivalDate: '2019-07-06T18:30:00.000Z',
            departureDate: '2019-07-07T18:30:00.000Z',
            hasShoulderDays: true,
            checkInShoulderDays: 3,
            checkOutShoulderDays: 3,
            allowRequestShoulderDays: true,
            hasPlannerApprovalShoulderDays: true,
            hasBillingInstructions: false,
            sourceType: 1,
            billingInstructions: [],
            billingInstructionOverrides: [],
            websiteUrl: '',
            distanceUnit: 'kilometers',
            roomTypes: [
              {
                id: '9a86e84a-9ca3-42f6-a0e3-e14df2ff2586',
                code: '',
                description: '',
                maxNumberOfRoomsPerRequest: 10,
                numberOfOccupantsPerRoom: 1,
                roomRateCalculationChoice: 1,
                sortOrder: 1,
                associatedRegPathSettings: {
                  '7121b5d7-604d-4e3f-9268-f85947d85337': {
                    regPathId: '7121b5d7-604d-4e3f-9268-f85947d85337',
                    allowCharge: false,
                    allowRefund: false
                  }
                },
                isMultiOccupancyRoom: false,
                isCombinedRoom: false,
                isOpenForRegistration: true,
                roomRate: [
                  {
                    rate: 80,
                    contactTypeId: '00000000-0000-0000-0000-000000000000',
                    capacityId:
                      '7503df79-64a7-4152-9315-4f1d27c87206::9a86e84a-9ca3-42f6-a0e3-e14df2ff2586::00000000-0000-0000-0000-000000000000::20190704',
                    isShoulderDate: true,
                    roomRateDate: '2019-07-03T18:30:00.000Z',
                    hasCapacity: true,
                    availableCapacity: 80
                  }
                ],
                roomTypeName: 'aa',
                minNumberOfNights: 1
              }
            ],
            sortOrder: 1
          }
        ],
        isHotelWebsiteEnabled: false,
        isHotelRequestEnabled,
        isPasskeyEnabled: false,
        hotelRequestRules: []
      }
    },

    travelCart: {
      cart: {
        bookings: [
          {
            hotelRequests: []
          }
        ]
      },
      userSession: {
        hotelRequest: {
          ownBooking: {}
        }
      }
    },
    text: {
      resolver: {
        date: () => 'some date ',
        currency: x => x
      },
      translate: x => x,
      translateDate: x => x
    },
    timezones: {
      35: {
        id: 35,
        name: 'Eastern Time',
        nameResourceKey: 'Event_Timezone_Name_35__resx',
        plannerDisplayName: '(GMT-05:00) Eastern [US & Canada]',
        hasDst: true,
        utcOffset: -300,
        abbreviation: 'ET',
        abbreviationResourceKey: 'Event_Timezone_Abbr_35__resx',
        dstInfo: []
      }
    },
    defaultUserSession: {
      isPlanner: false
    },
    appData: {
      registrationSettings: {
        registrationPaths: {
          'b8e50dbf-7437-4eec-a964-bc27d79d0373': {
            travelSettings: {
              hotelRequestSettings: {
                roommateRequestSettings: {
                  allowRequestFullRoom: true,
                  allowRoommatePreference: true,
                  roommatePreferenceOption: 'visible'
                }
              }
            }
          }
        }
      }
    },
    website: {
      theme: EventSnapshot.eventSnapshot.siteEditor.website.theme
    }
  };
}

async function dispatch(action) {
  if (typeof action === 'function') {
    await action(dispatch, getState);
  }
}

const subscribe = () => {};
const defaultProps = {
  classes: {},
  style: {},
  translate: c => c,
  store: { dispatch, getState, subscribe }
};

describe('EventHotelRequest', () => {
  test('should produce props from state', () => {
    const widget = shallow(<EventHotelRequest {...defaultProps} />);
    expect(widget.props()).toMatchSnapshot();
  });

  test('should not render when air request is not enabled', () => {
    isHotelRequestEnabled = false;
    const widget = shallow(<EventHotelRequest {...defaultProps} />);
    expect(widget).toMatchSnapshot();
    isHotelRequestEnabled = true;
  });

  test('should set allowHotelRequestDateChange to false for pending invitee from state', () => {
    const state = getState();
    const getPendingApprovalState = () => ({
      ...state,
      registrationForm: {
        ...state.registrationForm,
        regCart: {
          ...state.registrationForm.regCart,
          registrationApprovalRequired: true
        }
      }
    });
    const newProps = {
      classes: {},
      style: {},
      translate: c => c,
      store: { dispatch, getState: getPendingApprovalState, subscribe }
    };
    const widget = shallow(<EventHotelRequest {...newProps} />);
    expect(widget.props().allowHotelRequestDateChange).toBeFalsy();
  });
});
