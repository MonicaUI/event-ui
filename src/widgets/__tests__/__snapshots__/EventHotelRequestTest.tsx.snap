// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventHotelRequest should not render when air request is not enabled 1`] = `ShallowWrapper {}`;

exports[`EventHotelRequest should produce props from state 1`] = `
Object {
  "children": <HotelsRequestWidget
    accountHotelBillingInstructions={Array []}
    admissionItemId="3349dc98-e303-4bf9-b698-e1bdae85f946"
    attendeeRegTypeAndAdmItemIds={
      Array [
        "3349dc98-e303-4bf9-b698-e1bdae85f946",
      ]
    }
    classes={Object {}}
    derivedRoomRegTypeAssociations={Object {}}
    dialogContainerStyle={
      Object {
        "backgroundColor": "#F33B15",
        "borderRadius": 100,
        "hover": Object {
          "backgroundColor": "#F33B15",
        },
      }
    }
    displayPlannerFields={false}
    event={
      Object {
        "products": Object {
          "admissionItems": Object {
            "3349dc98-e303-4bf9-b698-e1bdae85f946": Object {
              "id": "3349dc98-e303-4bf9-b698-e1bdae85f946",
            },
          },
        },
        "registrationTypes": Object {
          "b8e50dbf-7437-4eec-a964-bc27d79d0372": Object {
            "id": "b8e50dbf-7437-4eec-a964-bc27d79d0372",
          },
        },
        "timezone": Object {
          "abbreviation": "ET",
          "abbreviationResourceKey": "Event_Timezone_Abbr_35__resx",
          "dstInfo": Array [],
          "hasDst": true,
          "id": 35,
          "name": "Eastern Time",
          "nameResourceKey": "Event_Timezone_Name_35__resx",
          "plannerDisplayName": "(GMT-05:00) Eastern [US & Canada]",
          "utcOffset": -300,
        },
      }
    }
    hotelRegRules={Array []}
    hotelsData={
      Object {
        "hotelRequestRules": Array [],
        "hotels": Array [
          Object {
            "address": "",
            "allowRequestShoulderDays": true,
            "arrivalDate": "2019-07-06T18:30:00.000Z",
            "billingInstructionOverrides": Array [],
            "billingInstructions": Array [],
            "cancellationPolicy": "",
            "checkInShoulderDays": 3,
            "checkOutShoulderDays": 3,
            "code": "",
            "departureDate": "2019-07-07T18:30:00.000Z",
            "description": "",
            "distanceUnit": "kilometers",
            "fax": "",
            "hasBillingInstructions": false,
            "hasPlannerApprovalShoulderDays": true,
            "hasShoulderDays": true,
            "hotelId": "9d4f772f-dd54-4e56-a436-d91aab911c9c",
            "id": "5ea37a1e-299a-4dc9-b1bf-c0ef74379fe2",
            "isActive": true,
            "name": "aa",
            "phone": "",
            "rateCode": "",
            "registrationLinkUrl": "",
            "roomTypes": Array [
              Object {
                "associatedRegPathSettings": Object {
                  "7121b5d7-604d-4e3f-9268-f85947d85337": Object {
                    "allowCharge": false,
                    "allowRefund": false,
                    "regPathId": "7121b5d7-604d-4e3f-9268-f85947d85337",
                  },
                },
                "code": "",
                "description": "",
                "id": "9a86e84a-9ca3-42f6-a0e3-e14df2ff2586",
                "isCombinedRoom": false,
                "isMultiOccupancyRoom": false,
                "isOpenForRegistration": true,
                "maxNumberOfRoomsPerRequest": 10,
                "minNumberOfNights": 1,
                "numberOfOccupantsPerRoom": 1,
                "roomRate": Array [
                  Object {
                    "availableCapacity": 80,
                    "capacityId": "7503df79-64a7-4152-9315-4f1d27c87206::9a86e84a-9ca3-42f6-a0e3-e14df2ff2586::00000000-0000-0000-0000-000000000000::20190704",
                    "contactTypeId": "00000000-0000-0000-0000-000000000000",
                    "hasCapacity": true,
                    "isShoulderDate": true,
                    "rate": 80,
                    "roomRateDate": "2019-07-03T18:30:00.000Z",
                  },
                ],
                "roomRateCalculationChoice": 1,
                "roomTypeName": "aa",
                "sortOrder": 1,
              },
            ],
            "sortOrder": 1,
            "sourceType": 1,
            "websiteLink": "",
            "websiteUrl": "",
          },
        ],
        "isHotelRequestEnabled": true,
        "isHotelWebsiteEnabled": false,
        "isOwnAccommodation": Object {},
        "isPasskeyEnabled": false,
        "selectedRooms": Array [],
      }
    }
    isGroupRegistration={false}
    isPendingInvitee={false}
    isPlanner={false}
    isRegMod={false}
    isRegistrationActionDisabled={false}
    isSiteEditorPreview={false}
    onClearRoomSelection={[Function]}
    onRequestShoulderDatesRequiringApproval={[Function]}
    onRoommateCriteriaChange={[Function]}
    onRoommateSearch={[Function]}
    onSubmitRoomRequest={[Function]}
    registrationPathId="b8e50dbf-7437-4eec-a964-bc27d79d0373"
    registrationTypeId="b8e50dbf-7437-4eec-a964-bc27d79d0372"
    registrations={
      Object {
        "guests": Array [],
        "primary": Object {
          "eventRegistrationId": "123456789",
          "productRegistrations": Array [
            Object {
              "productId": "3349dc98-e303-4bf9-b698-e1bdae85f946",
              "productType": "AdmissionItem",
              "requestedAction": "REGISTER",
            },
          ],
          "registrationPathId": "b8e50dbf-7437-4eec-a964-bc27d79d0373",
          "registrationTypeId": "b8e50dbf-7437-4eec-a964-bc27d79d0372",
        },
      }
    }
    roommateRequestSettings={
      Object {
        "allowRequestFullRoom": true,
        "allowRoommatePreference": true,
        "roommatePreferenceOption": "visible",
      }
    }
    setAnotherHotelRequestFlag={[Function]}
    showHotelRequestSummaryView={[Function]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "subscribe": [Function],
      }
    }
    style={Object {}}
    toggleOwnAccommodation={[Function]}
    translate={[Function]}
    translateCurrency={[Function]}
    translateDate={[Function]}
    updateExpandedHotels={[Function]}
  />,
  "value": null,
}
`;
